### ISSUE TITLE
"do while" loops

### ISSUE THREAD

**Issue by josepharhar** [whatwg/infra#567]
The current infra spec says you can have "while" and "for each" loops, but there are no "do while" loops: https://infra.spec.whatwg.org/#algorithm-iteration
I am trying to spec something that uses a "do while" loop, can we add do while loops here? Or is there another way to spec these?

@annevk @domenic

**Comment by domenic** [comment]
We could certainly add them. In the meantime I think you should feel free to use them in specs: the intent is pretty clear to any reader, and in general Infra is not meant to be a restriction on what you write, just a standard library so that if multiple people are writing the same thing, they do so the same way.

The spec-ese wording is a bit unclear to me: if you just wrote:

```
1. Do:

   1. X
   2. Y

  while Z.
```

the nature of English makes me think that this might just be a fancy way of writing

```
1. While Z, do:
   
   1. X
   2. Y
```

So maybe something like

```
1. Perform the following steps at least once:

   1. X
   2. Y
  
  and keep performing them while Z.
```

would be sufficiently clear.

I should also mention you can replace any do-while loop with a while loop, using a construct like

```
1. Let firstIteration be true.
2. While firstIteration is true or Z:
   1. Set firstIteration to false.
   2. X
   3. Y
```

which may be why languages like Python and Ruby don't have them.

**Comment by domenic** [comment]
https://stackoverflow.com/q/994905 is somewhat relevant.

**Comment by annevk** [comment]
I find "do while" loops a bit confusing so I'd prefer the explicit construct around "while", but I guess I could get used to them if they are defined and "Do" is xref'd.

**Comment by domenic** [comment]
@annevk , what'd you think of my more verbose phrasing ("...at least once... And keep performing them...")

**Comment by annevk** [comment]
Yeah, fair. Nit: s/the following/these/ and s/perform/run/ unless we want to switch everything from run to perform...
